### ISSUE TITLE
Could you clarify which async functionality is implemented?

### ISSUE THREAD

**Issue by sh1ng** [thrust/thrust#827]
Hi,

I'm a little confused which operations can be executed in async way and how to achieve it.
What I found 

- async reduce requires additional wrapper as global function or async with lambda https://github.com/thrust/thrust/blob/master/examples/cuda/async_reduce.cu

- for_each no need it https://github.com/thrust/thrust/blob/master/examples/cuda/simple_cuda_streams.cu

- There is an overload of copy function that accepts exec policy, so I can provide a stream. Will it result in async copy from host2device. Do I need to wrap that call in global function or async? I found only one  example on SO http://stackoverflow.com/questions/25064383/how-to-asynchronously-copy-memory-from-the-host-to-the-device-using-thrust-and-c it works with guaranty, but it uses a native api.

**Comment by egaburov** [comment]
* reduce, copy_if, transform_reduce, and other algorithms which return computed result are blocking. In part, because D2H copy is required before returning result to the calling thread.

* for_each is non-blocking. The return value can be, and is computed on the host, and no D2H copy is required before returning control to the calling thread. 

* If stream is provided via the execution policy, Thrust should synchronise in this stream only. H2D, D2D and D2H copies are implemented via cudaMemCopyAsync+sync in the provided stream. Thus copies are blocking.

If asynchronous execution is desired, the best bet would be to spawn a thread, e.g. via async, which will call Thrust algorithm. Please be mindful most of the Thrust algorithms allocate temporary storage. If you use multiple threads to invoke Thrust algorithms, it is best to provide a custom thread-safe allocator (e.g. [1]) via the execution policy (e.g. [2]).


[1] https://nvlabs.github.io/cub/structcub_1_1_caching_device_allocator.html
[2] https://github.com/thrust/thrust/blob/master/examples/cuda/custom_temporary_allocation.cu

**Comment by sh1ng** [comment]
@egaburov  thank so much for detailed answers. In my case all operation don't return result, so they should work in async manner. 

According to http://stackoverflow.com/questions/14093601/effect-of-using-page-able-memory-for-asynchronous-memory-copy I also should use pinned memory to get full async behavior. Is it correct? 
There is a experimental pinned_allocator. Are you going to continue maintain it?

I'm asking so much questions because it's not clear is it worth to use thrust with all its limitations(and inactive development) or focus on native api.

**Comment by egaburov** [comment]
I do not think Thrust spells out guarantees of async behaviour, hence "should work" doesn't not necessarily imply "will work". If you absolutely must have async behaviour, perhaps spawning a separate thread is what you need.

cudaMemcpyAsync requires pinned host memory buffer for non-blocking operation. thrust::copy, however, is a blocking call, whether the host buffer is pinned or not.

Hope this helps.

**Comment by sh1ng** [comment]
I'm going to use a trick from http://stackoverflow.com/questions/25064383/how-to-asynchronously-copy-memory-from-the-host-to-the-device-using-thrust-and-c

**Comment by sh1ng** [comment]
or try experimental pinned_allocator

**Comment by sh1ng** [comment]
@egaburov according documentation thrust:for_each returns  InputIterator. So it should block current thread(and only a few methods in Thrust return void). Does it mean all of them are blocking?

**Comment by egaburov** [comment]
Let me rephrase it differently. Thrust doesn't spell out guarantees of non-blocking behaviour, and its current interface assumes blocking operation. for_each, and I think the other for_each-based algorithms, is an exception rather to this. It is not guaranteed that they will stay non-blocking.

Hope this helps.

**Comment by sh1ng** [comment]
I've finally decided to use cub. It gives more control.

Спасибо за ответы.

**Comment by egaburov** [comment]
We are looking for user feedback to improve Thrust. May I ask you to elaborate with kind of control CUB provides you, and what would you like to have in Thrust to accomplish your tasks? 

Please feel to file an RFE, either via GitHub issue or by submitting a bug via formal NVIDIA bug tracking system, to tell us what  unctionality would you like to see in Thrust.

Thank you.

**Comment by sh1ng** [comment]
First of all I don't have a lot experience with cuda and thrust, so my opinions could be really naive.

- First of all it's async behavior(to mix memory copy and computation).  As you proposed I can spawn a thread and operation will block that thread, but my code becomes over complicated. I can do the same with cuda stream(s) and stream sync. 

- A lot of algorithms use temp data. It's worth to reuse it to eliminate cudaMaloc, cudaFree. Cub allows me to do it explicitly or with caching device allocate.

- More control on algorithm. like begin_bit, end_bit for Radix sort http://nvlabs.github.io/cub/structcub_1_1_device_radix_sort.html#a31d7224d3f6c6a9309a0b84571f25eb9 

- More general observation. Sometimes cuda effective implementation is 'hacky' and takes into account tiny little details specific for a certain architecture. It's good to have template framework like Thrust as a starting point when you learn cuda, but when you go deeper it's easier to (at least for me) learn some native mechanism(streams, async memory copy, etc) and cub don't restrict you with that. Thrust in contrast is to abstract(I know you have to support cuda, openmp, etc) and not so transparent as cub is.

**Comment by egaburov** [comment]
Thank you for your feedback.

> A lot of algorithms use temp data. It's worth to reuse it to eliminate cudaMaloc, cudaFree. Cub allows me to do it explicitly or with caching device allocate.

Have you considered using custom allocators which can supplied via execution policy [1]

> More control on algorithm. like begin_bit, end_bit for Radix sort

This is only makes sense for radix sort. But I suppose, what you imply is that it would be good if Thrust,  in addition to generic comparison sorts thrust::sort, thrust::stable_sort, would also have thrust::radix_sort with an optional argument to control bit range.


[1] https://github.com/thrust/thrust/blob/master/examples/cuda/custom_temporary_allocation.cu

**Comment by sh1ng** [comment]
`Have you considered using custom allocators which can supplied via execution policy [1]`
Yeap, but in my case algorithm looks like
[h2d copy][allocate tmp data][computation][d2h copy]
If I do it in parallel 
[h2d copy][allocate tmp data][computation][d2h copy]
                 [h2d copy][allocate tmp data][computation][d2h copy]
                                  [h2d copy][allocate tmp data][computation][d2h copy]
                                                   [h2d copy][allocate tmp data][computation][d2h copy]

it's more efficient to allocate continuous peace of memory for a `temp data` when you start and reuse it. So I don't event need a cub::CachingDeviceAllocator. And cub allows just provide pointer for a temp storage. For me it's easier and more readable than implement custom allocator.

`This is only makes sense for radix sort. But I suppose, what you imply is that it would be good if Thrust, in addition to generic comparison sorts thrust::sort, thrust::stable_sort, would also have thrust::radix_sort with an optional argument to control bit range.`

I will rephrase, a lot of algorithms cab be configured to leverage performance by using  architecture specific features. cub::BlockHistogram could be done by BLOCK_HISTO_SORT  or BLOCK_HISTO_ATOMIC for instance. It's really tremendous amount of work to support multiple device architecture(cuda, openmp, etc) and the same level of native features. Might be you should consider to focus on friction-less usage of low level libraries(like cub) inside Thrust. That will help developers to get desirable level of control and performance but keep STL like code.
